test/base/vault/DiamondVaultTest.ts

```diff
@@ -0,0 +1,420 @@
+import {ethers} from "hardhat";
+import chai from "chai";
+import {
+  ContractReader,
+  Multicall,
+  MultiSwap,
+  NoopStrategy,
+  SmartVault,
+  ZapContract
+} from "../../../typechain";
+import {DeployerUtils} from "../../../scripts/deploy/DeployerUtils";
+import {VaultUtils} from "../../VaultUtils";
+import {utils} from "ethers";
+import {Erc20Utils} from "../../Erc20Utils";
+import {TimeUtils} from "../../TimeUtils";
+import {SignerWithAddress} from "@nomiclabs/hardhat-ethers/signers";
+import chaiAsPromised from "chai-as-promised";
+import {CoreContractsWrapper} from "../../CoreContractsWrapper";
+import {MaticAddresses} from "../../MaticAddresses";
+import {UniswapUtils} from "../../UniswapUtils";
+import {MintHelperUtils} from "../../MintHelperUtils";
+
+const {expect} = chai;
+chai.use(chaiAsPromised);
+
+const LOCK_DURATION = 60 * 60 * 24 * 3;
+const LOCK_PENALTY = 400;
+
+describe("Diamond vault test", () => {
+  let snapshot: string;
+  let snapshotForEach: string;
+  let signer: SignerWithAddress;
+  let core: CoreContractsWrapper;
+  let contractReader: ContractReader;
+  let zapContract: ZapContract;
+  let multiSwap: MultiSwap;
+  let multicall: Multicall;
+
+  before(async function () {
+    snapshot = await TimeUtils.snapshot();
+    signer = (await ethers.getSigners())[0];
+    core = await DeployerUtils.deployAllCoreContracts(signer);
+
+    const calculator = (await DeployerUtils.deployPriceCalculatorMatic(signer, core.controller.address))[0];
+
+    multicall = await DeployerUtils.deployContract(signer, "Multicall") as Multicall;
+    const crLogic = await DeployerUtils.deployContract(signer, "ContractReader");
+    const crProxy = await DeployerUtils.deployContract(signer, "TetuProxyGov", crLogic.address);
+    contractReader = crLogic.attach(crProxy.address) as ContractReader;
+
+    await contractReader.initialize(core.controller.address, calculator.address);
+
+    multiSwap = await DeployerUtils.deployMultiSwap(signer, core.controller.address, calculator.address);
+    zapContract = (await DeployerUtils.deployZapContract(signer, core.controller.address, multiSwap.address));
+    await core.controller.addToWhiteList(zapContract.address);
+
+    await UniswapUtils.buyAllBigTokens(signer);
+  });
+
+  after(async function () {
+    await TimeUtils.rollback(snapshot);
+  });
+
+  beforeEach(async function () {
+    snapshotForEach = await TimeUtils.snapshot();
+  });
+
+  afterEach(async function () {
+    await TimeUtils.rollback(snapshotForEach);
+  });
+
+
+  it("check locked funds in loop", async () => {
```

Comment:
This test is too generic, I can't pick up testing scenarios. Could you please divide it for small isolated use cases.

Reply:
some functionality can't be tested without multiple interactions with the vault
it requires a loop with users actions that probably can affect some logic